File: BOJ/Kruskal/1197-최소 스패닝 트리.java
import java.util.*;
import java.io.*;

class Main {
    static int v, e;
    static int[] parents;
    static ArrayList<int[]> graph = new ArrayList<>();

    public static int find(int a) {
        if (parents[a] != a) {
            parents[a] = find(parents[a]);
        }
        return parents[a];
    }

    public static void union(int a, int b) {
        a = find(a);
        b = find(b);

        if (a > b) {
            parents[b] = a;
        } else {
            parents[a] = b;
        }
    }

    public static int kruscal() {
        Collections.sort(graph, (o1, o2) -> Integer.compare(o1[0], o2[0]));
        int r = 0, cnt = 0;
        for (int[] tmp : graph) {
            int c = tmp [0];
            int a = tmp[1];
            int b = tmp[2];

            if (find(a) != find(b)) {
                union(a, b);
                r += c;
                cnt++;
                if (cnt == v - 1) break;
            }
        }
        return r;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        v = Integer.parseInt(st.nextToken());
        e = Integer.parseInt(st.nextToken());

        parents = new int[v + 1];
        for (int i = 0; i <= v; i++) {
            parents[i] = i;
        }

        for (int i = 0; i < e; i++) {
            int a, b, c;
            st = new StringTokenizer(br.readLine());
            a = Integer.parseInt(st.nextToken());
            b = Integer.parseInt(st.nextToken());
            c = Integer.parseInt(st.nextToken());

            graph.add(new int[]{c, a, b});
        }

        System.out.println(kruscal());
    }
}

File: BOJ/Kruskal/1197-최소 스패닝 트리.py
import sys

input = sys.stdin.readline

def find(parents, a):
    if parents[a] != a:
        parents[a] = find(parents, parents[a])
    return parents[a]

def union(parents, a, b):
    a = find(parents, a)
    b = find(parents, b)

    if a > b:
        parents[b] = a
    else:
        parents[a] = b

def kruskcal(parents, graph):
    total = 0
    graph.sort()
    # graph.sort(lambda x : x[0])
    for c, a, b in graph:
        if find(parents, a) != find(parents, b):
            union(parents, a, b)
            total += c
    return total

v, e = map(int, input().split())
graph = []
parents = [i for i in range(v + 1)]

for _ in range(e):
    a, b, c = map(int, input().split())
    graph.append([c, a, b])

print(kruskcal(parents, graph))